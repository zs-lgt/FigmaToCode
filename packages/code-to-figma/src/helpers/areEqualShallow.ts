export const areEqualShallow = (obj1?: Object, obj2?: Object): boolean => {
    if (Object.is(obj1, obj2)) {
        return true;
    }

    if (typeof obj1 !== 'object' || obj1 === null || typeof obj2 !== 'object' || obj2 === null) {
        return false;
    }

    return (
        Object.keys(obj1).length === Object.keys(obj2).length &&
        Object.keys(obj1).every(key => obj2.hasOwnProperty(key) && obj1[key] === obj2[key])
    );
};
