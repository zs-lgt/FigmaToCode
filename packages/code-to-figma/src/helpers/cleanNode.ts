export const cleanNode = node => {
    if (!node) {
        return;
    }
    const { children, parent, ref, innerRef, ...other } = node;
    return other;
};
