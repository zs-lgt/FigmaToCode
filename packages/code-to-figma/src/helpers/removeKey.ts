export const removeKey = node => {
    if (node && node.key) {
        delete node.key;
    }
    if (node && node.children) {
        node.children.forEach(removeKey);
    }
    return node;
};
