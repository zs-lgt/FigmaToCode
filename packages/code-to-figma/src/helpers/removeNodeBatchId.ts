export const removeNodeBatchId = node => {
    if (node && node.pluginData) {
        delete node.pluginData.nodeBatchId;
    }
    if (node && node.children) {
        node.children.forEach(removeNodeBatchId);
    }
    return node;
};
