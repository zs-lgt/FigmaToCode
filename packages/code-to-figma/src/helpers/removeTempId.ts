export const removeTempId = node => {
    if (node && node.pluginData) {
        delete node.pluginData.reactId;
    }
    if (node && node.children) {
        node.children.forEach(removeTempId);
    }
    return node;
};
