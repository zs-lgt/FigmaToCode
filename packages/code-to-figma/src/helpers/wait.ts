import { act } from 'react-test-renderer';

export function wait(amount = 0) {
    return new Promise(resolve => setTimeout(resolve, amount));
}

// Use this in your test after mounting if you need just need to let the query finish without updating the wrapper
export async function actWait(amount = 0) {
    await act(async () => {
        await wait(amount);
    });
}
