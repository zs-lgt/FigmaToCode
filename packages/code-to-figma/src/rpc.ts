import { createPluginAPI, createUIAPI } from 'figma-jsonrpc';
import { isReactFigmaNode } from './helpers/isReactFigmaNode';
import * as renderers from './renderers';
import { nanoid } from 'nanoid/non-secure';
import { Subject } from 'rxjs';
import { safeGetPluginData } from './helpers/safeGetPluginData';
import { LayoutStyleProperties } from './styleTransformers/transformLayoutStyleProperties';
import { GeometryStyleProperties } from './styleTransformers/transformGeometryStyleProperties';
import { isEqualFontStyle } from './helpers/isEqualFontStyle';
import { CommonStyleProps } from './types';
import { DEFAULT_FONT } from './helpers/constants';

const getInitialTree = node => {
    return {
        id: node.id,
        type: node.type,
        reactId: safeGetPluginData('reactId')(node),
        children:
            node.children && node.children.filter(item => isReactFigmaNode(item)).map(item => getInitialTree(item))
    };
};

const findRoot = (node: any) => {
    if (!node) {
        return;
    }
    const parent = node.parent;
    if (!parent || !isReactFigmaNode(parent)) {
        return node;
    } else {
        return findRoot(parent);
    }
};

const transformNodesToTree = node => {
    if (!isReactFigmaNode(node)) {
        return;
    }
    const nodeBatchId = nanoid();
    node.setPluginData('nodeBatchId', nodeBatchId);
    const children = node.children && node.children.map(transformNodesToTree).filter(item => !!item);
    return {
        width: node.width,
        height: node.height,
        style:
            (safeGetPluginData('reactStyle')(node) && JSON.parse(safeGetPluginData('reactStyle')(node) || '')) ||
            undefined,
        children: children && children.length > 0 ? children : undefined,
        reactId: safeGetPluginData('reactId')(node),
        nodeBatchId
    };
};

const renderInstance = (type, node, props, reactId) => {
    const instance = renderers[type](node)(props);
    if (!node) {
        instance.setPluginData('isReactFigmaNode', 'true');
        instance.setPluginData('reactId', reactId);
    }
    return instance;
};

const cleanStubElements = parentNode => {
    if (parentNode.type === 'GROUP') {
        parentNode.children.forEach(child => {
            if (safeGetPluginData('isGroupStubElement')(child)) {
                child.remove();
            }
        });
    }

    if (parentNode.type === 'COMPONENT_SET') {
        parentNode.children.forEach(child => {
            if (safeGetPluginData('isComponentStubElement')(child)) {
                child.remove();
            }
        });
    }
};

const appendToContainer = (parentNode, childNode) => {
    if (!childNode || !parentNode || parentNode.type === 'INSTANCE') {
        return;
    }

    parentNode.appendChild(childNode);
    cleanStubElements(parentNode);
};

const insertToContainer = (parentNode, newChildNode, beforeChildNode) => {
    if (!parentNode || !newChildNode || !beforeChildNode || parentNode.type === 'INSTANCE') {
        return;
    }
    const beforeChildIndex = parentNode.children.indexOf(beforeChildNode);
    parentNode.insertChild(beforeChildIndex, newChildNode);
    cleanStubElements(parentNode);
};

const cache = {};

const transformToNode = smth => {
    if (!smth) {
        return;
    }
    if (smth.id) {
        return figma.getNodeById(smth.id);
    } else if (smth.reactId) {
        return cache[smth.reactId];
    } else {
        return smth;
    }
};

const findNodeByName = (children, name) => {
    return children && children.find(child => child.name === name || findNodeByName(child.children, name));
};

export const api = createPluginAPI(
    {
        getInitialTree() {
            return getInitialTree(figma.root);
        },

        renderInstance(type, _node, props, tempNode) {
            const node = transformToNode(_node);
            const instance = renderInstance(
                type,
                node,
                props && {
                    ...props,
                    ...(type === 'instance' && props.component ? { component: transformToNode(props.component) } : {}),
                    ...(props.node ? { node: transformToNode(props.node) } : {})
                },
                tempNode.reactId
            );
            cache[tempNode.reactId] = instance;
            uiApi.bindReactIdWithNode(tempNode.reactId, instance.id);
        },

        appendToContainer(_parentNode, _childNode) {
            const parentNode = transformToNode(_parentNode);
            const childNode = transformToNode(_childNode);
            appendToContainer(parentNode, childNode);
        },

        insertToContainer(_parentNode, _newChildNode, _beforeChildNode) {
            const parentNode = transformToNode(_parentNode);
            const newChildNode = transformToNode(_newChildNode);
            const beforeChildNode = transformToNode(_beforeChildNode);
            insertToContainer(parentNode, newChildNode, beforeChildNode);
        },

        async listAvailableFontsAsync() {
            return figma.listAvailableFontsAsync();
        },

        async loadFontAsync(fontName) {
            return figma.loadFontAsync(fontName);
        },

        remove(_childNode) {
            const childNode = transformToNode(_childNode);
            if (!childNode || childNode.removed) {
                return;
            }
            childNode.remove();
        },

        getTreeForYoga(_instance) {
            const node = transformToNode(_instance);
            const root = findRoot(node);
            return transformNodesToTree(root);
        },

        findNodeByName(_node, name) {
            const node = transformToNode(_node);
            const instanceItemNode = findNodeByName(node.children, name);
            return instanceItemNode && getInitialTree(instanceItemNode);
        },

        createImage(data) {
            const image = figma.createImage(data);
            return image.hash;
        },

        setCurrentPage(_node) {
            const node = transformToNode(_node);
            figma.currentPage = node;
        },

        highlightNativeElement(_node) {
            const node = transformToNode(_node);

            if (!node || node.type === 'DOCUMENT') {
                return;
            }

            if (figma.currentPage.selection.includes(node)) {
                return;
            }

            const nodePage = findRoot(node);

            if (figma.currentPage !== nodePage) {
                figma.currentPage = nodePage;
            }

            if (node.type !== 'PAGE') {
                figma.viewport.scrollAndZoomIntoView([node]);
                figma.currentPage.selection = [node];
            }
        },

        async importStyleByKeyAsync(key: string): Promise<BaseStyle> {
            return figma.importStyleByKeyAsync(key);
        },

        createOrUpdatePaintStyle(properties: {
            paints: ReadonlyArray<Paint> | symbol | void;
            params: CommonStyleProps;
        }) {
            const { paints, params } = properties;
            const { name, id, description } = params;
            const foundPaintStyle = figma.getLocalPaintStyles().find(style => style.name === name || style.id === id);
            const paintStyle = foundPaintStyle || figma.createPaintStyle();
            if (name) {
                paintStyle.name = name;
            }
            if (description) {
                paintStyle.description = description;
            }
            if (paints) {
                paintStyle.paints = paints as any;
            }
            return paintStyle.id;
        },

        createOrUpdateTextStyle(properties: { textProperties: any | void; params: CommonStyleProps; loadedFont: any }) {
            const { textProperties = {}, params, loadedFont } = properties;
            const { name, id, description } = params;
            const foundTextStyle = figma.getLocalTextStyles().find(style => style.name === name || style.id === id);
            const textStyle = foundTextStyle || figma.createTextStyle();
            if (name) {
                textStyle.name = name;
            }
            if (description) {
                textStyle.description = description;
            }

            const { fontName = DEFAULT_FONT } = textProperties;
            if (
                loadedFont &&
                fontName &&
                loadedFont.family === fontName.family &&
                isEqualFontStyle(loadedFont.style, fontName.style)
            ) {
                Object.keys(textProperties).forEach(key => {
                    textStyle[key] = textProperties[key];
                });
            }
            return textStyle.id;
        },

        createOrUpdateEffectStyle(properties: {
            effects: ReadonlyArray<Effect> | symbol | void;
            params: CommonStyleProps;
        }) {
            const { effects, params } = properties;
            const { name, id, description } = params;
            const foundEffectStyle = figma.getLocalEffectStyles().find(style => style.name === name || style.id === id);
            const effectStyle = foundEffectStyle || figma.createEffectStyle();
            if (name) {
                effectStyle.name = name;
            }
            if (description) {
                effectStyle.description = description;
            }
            if (effects) {
                effectStyle.effects = effects as any;
            }
            return effectStyle.id;
        }
    },
    {
        timeout: 60 * 1000
    }
);

export const setupMainThread = () => {
    figma.on('currentpagechange', () => {
        const reactId = figma.currentPage.getPluginData('reactId');
        uiApi.currentPageChange(reactId);
    });

    figma.on('selectionchange', () => {
        const reactIds = figma.currentPage.selection.map(safeGetPluginData('reactId'));
        uiApi.selectionChange(reactIds);
    });
};

export const $currentPageTempId = new Subject();

export const $selectionReactIds = new Subject();

export const $updateYogaReactId = new Subject();

export const $bindReactIdWithNodeId = new Subject<[string, string]>();

// those methods will be executed in the Figma UI,
// regardless of where they are called from
export const uiApi = createUIAPI(
    {
        currentPageChange: reactId => {
            $currentPageTempId.next(reactId);
        },
        selectionChange: reactIds => {
            $selectionReactIds.next(reactIds);
        },
        updateYogaNode: reactId => {
            $updateYogaReactId.next(reactId);
        },
        bindReactIdWithNode: (reactId, nodeId) => {
            $bindReactIdWithNodeId.next([reactId, nodeId]);
        }
    },
    {
        timeout: 60 * 1000
    }
);
