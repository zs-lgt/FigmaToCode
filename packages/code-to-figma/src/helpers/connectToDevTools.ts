import { connectToDevTools as connectToDevToolsInternal } from 'react-devtools-core';
import { api } from '../rpc';

export const connectToDevTools = () => {
    let connectionFailed = false;
    const connection = new WebSocket('ws://localhost:8097');

    connectToDevToolsInternal({
        // prevents from multiple retries
        isAppActive: () => !connectionFailed,
        resolveRNStyle: null,
        websocket: connection
    });

    connection.onerror = () => {
        console.warn('React DevTools is not running. Please run React DevTools or remove `connectToDevTools` call');
        connectionFailed = true;
    };

    // @ts-ignore
    window.__REACT_DEVTOOLS_GLOBAL_HOOK__.on('react-devtools', agent => {
        agent._bridge.addListener('inspectElement', data => {
            const { rendererID, id } = data;

            const renderer = agent.rendererInterfaces[rendererID];

            let nodes = null;
            if (renderer !== null) {
                nodes = renderer.findNativeNodesForFiberID(id);
            }

            if (nodes !== null && nodes[0]) {
                api.highlightNativeElement(nodes[0]);
            }
        });
    });
};
