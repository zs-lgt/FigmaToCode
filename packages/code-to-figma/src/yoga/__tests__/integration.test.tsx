import * as React from 'react';
import { createFigma } from 'figma-api-stub';
import { Page, render, View } from '../..';
import { wait } from '../../helpers/wait';
import { removeNodeBatchId } from '../../helpers/removeNodeBatchId';
import { removeTempId } from '../../helpers/removeTempId';
import '../../rpc';
import { removeKey } from '../../helpers/removeKey';

const removeMeta = node => {
    return removeKey(removeNodeBatchId(removeTempId(node)));
};

describe('Yoga layout integration', () => {
    beforeEach(() => {
        // @ts-ignore
        global.figma = createFigma({
            simulateErrors: true
        });
    });

    it('render single component', async () => {
        await render(
            <Page>
                <View>
                    <View style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                    <View style={{ width: 200, height: 100, backgroundColor: '#ff0017' }} />
                </View>
            </Page>
        );
        await wait();
        await wait();
        expect(removeMeta(figma.root)).toMatchSnapshot();
        expect(figma.root.children['1'].children['0'].height).toEqual(200);
        expect(figma.root.children['1'].children['0'].children['1'].y).toEqual(100);
    });
});
