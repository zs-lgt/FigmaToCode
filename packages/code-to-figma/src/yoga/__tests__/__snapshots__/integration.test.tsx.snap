// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Yoga layout integration render single component 1`] = `
ChildrenMixinStub {
  "children": Array [
    ExportMixinStub {
      "children": Array [],
      "config": undefined,
      "id": "0:1",
      "parent": [Circular],
      "type": "PAGE",
    },
    ExportMixinStub {
      "_backgrounds": Array [
        Object {
          "blendMode": "NORMAL",
          "color": Object {
            "b": 0.9607843160629272,
            "g": 0.9607843160629272,
            "r": 0.9607843160629272,
          },
          "opacity": 1,
          "type": "SOLID",
          "visible": true,
        },
      ],
      "children": Array [
        GeometryMixinStub {
          "_fills": Array [],
          "backgrounds": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0,
                "r": 0,
              },
              "opacity": 0,
              "type": "SOLID",
            },
          ],
          "blendMode": "NORMAL",
          "children": Array [
            GeometryMixinStub {
              "_fills": Array [
                Object {
                  "color": Object {
                    "b": 0,
                    "g": 1,
                    "r": 0.07058823529411765,
                  },
                  "opacity": 1,
                  "type": "SOLID",
                },
              ],
              "blendMode": "NORMAL",
              "config": Object {
                "isWithoutTimeout": false,
                "simulateErrors": true,
              },
              "constraints": Object {
                "horizontal": "MIN",
                "vertical": "MIN",
              },
              "cornerRadius": 0,
              "cornerSmoothing": 0,
              "effects": Array [],
              "exportSettings": Array [],
              "height": 100,
              "id": "1:2",
              "isMask": false,
              "layoutAlign": "INHERIT",
              "layoutGrow": 0,
              "locked": false,
              "opacity": 1,
              "parent": [Circular],
              "pluginData": Object {
                "isReactFigmaNode": "true",
                "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
              },
              "strokeAlign": "INSIDE",
              "strokeCap": "NONE",
              "strokeWeight": 0,
              "strokes": Array [],
              "type": "RECTANGLE",
              "visible": true,
              "width": 200,
              "x": 0,
              "y": 0,
            },
            GeometryMixinStub {
              "_fills": Array [
                Object {
                  "color": Object {
                    "b": 0.09019607843137255,
                    "g": 0,
                    "r": 1,
                  },
                  "opacity": 1,
                  "type": "SOLID",
                },
              ],
              "blendMode": "NORMAL",
              "config": Object {
                "isWithoutTimeout": false,
                "simulateErrors": true,
              },
              "constraints": Object {
                "horizontal": "MIN",
                "vertical": "MIN",
              },
              "cornerRadius": 0,
              "cornerSmoothing": 0,
              "effects": Array [],
              "exportSettings": Array [],
              "height": 100,
              "id": "1:3",
              "isMask": false,
              "layoutAlign": "INHERIT",
              "layoutGrow": 0,
              "locked": false,
              "opacity": 1,
              "parent": [Circular],
              "pluginData": Object {
                "isReactFigmaNode": "true",
                "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#ff0017\\"}",
              },
              "strokeAlign": "INSIDE",
              "strokeCap": "NONE",
              "strokeWeight": 0,
              "strokes": Array [],
              "type": "RECTANGLE",
              "visible": true,
              "width": 200,
              "x": 0,
              "y": 100,
            },
          ],
          "clipsContent": false,
          "config": Object {
            "isWithoutTimeout": false,
            "simulateErrors": true,
          },
          "constraints": Object {
            "horizontal": "MIN",
            "vertical": "MIN",
          },
          "cornerRadius": 0,
          "cornerSmoothing": 0,
          "counterAxisAlignItems": "MIN",
          "counterAxisSizingMode": "AUTO",
          "effects": Array [],
          "exportSettings": Array [],
          "guides": Array [],
          "height": 200,
          "id": "1:4",
          "isMask": false,
          "itemSpacing": 0,
          "layoutAlign": "INHERIT",
          "layoutGrids": Array [],
          "layoutGrow": 0,
          "layoutMode": "NONE",
          "locked": false,
          "opacity": 1,
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"backgroundColor\\":\\"transparent\\"}",
          },
          "primaryAxisAlignItems": "MIN",
          "primaryAxisSizingMode": "AUTO",
          "strokeAlign": "INSIDE",
          "strokeCap": "NONE",
          "strokeWeight": 0,
          "strokes": Array [],
          "type": "FRAME",
          "visible": true,
          "width": 200,
          "x": 0,
          "y": 0,
        },
      ],
      "config": Object {
        "isWithoutTimeout": false,
        "simulateErrors": true,
      },
      "exportSettings": Array [],
      "id": "1:1",
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
        "reactStyle": "{}",
      },
      "type": "PAGE",
    },
  ],
  "config": undefined,
  "id": "0:0",
  "type": "DOCUMENT",
}
`;
